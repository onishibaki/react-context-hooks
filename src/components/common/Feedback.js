import React, { useState } from "react";
import { useEmployeeState } from "../context";
import styled from "styled-components";

import {
  ActionTypes,
  updateFeedback,
  addFeedback,
  updateEmployeeFlag,
} from "../action";

const FeedBackContainer = styled.div`
  border: 1px solid black;
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  text-align: center;
`;

const FeedBackWrapper = styled.div`
  border: 1px solid rgb(17, 17, 17);
  background-color: white;
  margin-top: 158px;
  display: inline-block;
  padding: 35px 10px 10px 10px;
  position: relative;
`;

const SpanButton = styled.span`
  font-weight: 700;
  color: black;
  font-size: 15px;
  margin: 10px 0;
  display: inline-block;
  position: absolute;
  top: -8px;
  right: 9px;
  cursor: pointer;
}
`;

const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 10px 0;
  display: block;
  &:hover {
    background: #00000059;
    color: white;
  }
`;

const Textarea = styled.textarea`
  display: block;
  height: 100px;
  width: 350px;
`;

const Label = styled.label`
  display: block;
  text-align: left;
`;

export const Feedback = () => {
  //Use Context
  const [
    {
      formValue,
      employeeId,
      feedbackValue: { feedback_id: feedBackId, feedback },
      formMessage,
    },
    dispatch,
  ] = useEmployeeState();
  //Set state for FeedBack Changes
  const [stateFeedback, setFeedback] = useState("");

  //Checking for State if feedBack name is blank to throw error
  const [stateUpdateFeedback, setUpdateFeedback] = useState(null);

  //Set the First Value
  let feedBackParameter = feedback;

  //Parameter for Save
  const feedBackSaveParam = {
    employeeId,
    stateFeedback,
    flag: stateFeedback ? 0 : 1,
  };

  //Checking for State if feedBack is blank to throw error
  if (stateUpdateFeedback === "") {
    feedBackParameter = stateUpdateFeedback;
  }

  //Parameter for Update
  const feedBackUpdateParam = {
    employeeId,
    feedBackId,
    stateFeedback: stateFeedback || feedBackParameter,
  };

  //Getting the value of Input changes
  const inputChangedHandler = (event) => {
    setFeedback(event.target.value);
    setUpdateFeedback("");
  };

  return (
    //Condition to display Feedback Screen
    formValue !== "" && (
      <FeedBackContainer className="feedback-container">
        <FeedBackWrapper>
          <SpanButton
            onClick={() => {
              dispatch({
                type: ActionTypes.FORM_MESSAGE,
                data: "",
              });
              dispatch({
                type: ActionTypes.SET_FORM_STATE_FEEDBACK,
                form: "",
                data: {},
                id: "",
              });
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              });
              setFeedback("");
              setUpdateFeedback(null);
            }}
          >
            X
          </SpanButton>
          <Label>Feedback: </Label>
          <Textarea
            name="feedback"
            defaultValue={feedback}
            onChange={inputChangedHandler}
          ></Textarea>
          {Object.keys(formMessage).length !== 0 && (
            <h2>{formMessage.message}</h2>
          )}
          {formValue !== "update" ? (
            <Button
              onClick={() => {
                addFeedback(feedBackSaveParam).then(function (response) {
                  dispatch(response);
                });
                updateEmployeeFlag(feedBackSaveParam);
              }}
            >
              Save
            </Button>
          ) : (
            <Button
              onClick={() => {
                updateFeedback(feedBackUpdateParam).then(function (response) {
                  dispatch(response);
                });
              }}
            >
              Update
            </Button>
          )}
        </FeedBackWrapper>
      </FeedBackContainer>
    )
  );
};
