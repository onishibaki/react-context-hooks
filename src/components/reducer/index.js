import React, { useReducer } from "react";

import { ActionTypes } from "../action";
import { EmployeesContext } from "../context";

const appStateReducer = (state, action) => {
  switch (action.type) {
    case ActionTypes.GET_ALL_EMPLOYEE:
      return {
        ...state,
        employees: action.data,
      };
    case ActionTypes.FORM_MESSAGE:
      return {
        ...state,
        formMessage: action.data,
      };
    case ActionTypes.SET_FORM_STATE:
      return {
        ...state,
        form: action.form,
        employee: action.data,
      };
    case ActionTypes.SET_FORM_STATE_FEEDBACK:
      return {
        ...state,
        formValue: action.form,
        feedbackValue: action.data,
        employeeId: action.id,
      };
    default:
      return state;
  }
};

const initialState = {
  employees: {},
  form: "",
  employee: {},
  formValue: "",
  feedbackValue: {},
  id: "",
  formMessage: "",
};

export const AppEmployeesProvider = ({ children }) => {
  const employeesList = useReducer(appStateReducer, initialState);
  return (
    <EmployeesContext.Provider value={employeesList}>
      {children}
    </EmployeesContext.Provider>
  );
};
