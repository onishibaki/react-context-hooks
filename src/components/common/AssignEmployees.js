import React from "react";
import { useEmployeeState } from "../context";
import styled from "styled-components";

import { ActionTypes, GetAllEmployees } from "../action";
import { Feedback } from "./Feedback";

const EmployeesWrapper = styled.div`
  border: 1px solid #000000;
  margin-top: 30px;
`;
const EmployeesDetailWrapper = styled.div`
  width: 60%;
`;

const TopMenu = styled.div`
  margin: 22px 10px;
  a {
    text-decoration: none;
    border: 1px solid #000000;
    padding: 10px 20px;
    font-weight: 700;
    color: black;
    font-size: 12px;
    margin: 0 20px;
    &:hover {
      background: #00000059;
      color: white;
    }
  }
`;

const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 0 20px;
  &:hover {
    background: #00000059;
    color: white;
  }
`;

const Unordered = styled.ul`
  list-style: none;
  padding: 0;
  display: flex;
  justify-content: space-around;
  flex-direction: column;
  padding: 0 20px;
  li {
    display: flex;
    margin: 5px 0;
  }
`;
const Span = styled.span`
  font-size: 15px;
  padding: 0 20px;
  font-weight: 700;
`;

export const AssignEmployeeList = ({ props }) => {
  const { employees, status, message } = props;
  // eslint-disable-next-line
  const [stateDispatch, dispatch] = useEmployeeState();
  GetAllEmployees().then(function (response) {
    dispatch(response);
  });

  const empFlagFilter = employees.filter(
    (employeeList) => employeeList.flag === "1"
  );

  return status === 0 ? (
    <div className="main-container">{message}</div>
  ) : (
    <div className="main-container">
      <EmployeesWrapper>
        <TopMenu>
          <span>
            <a href="/">Home</a>
          </span>
        </TopMenu>
        <Unordered>
          {Object.keys(empFlagFilter).length === 0 && (
            <h5>No Assigned Employees</h5>
          )}
          {empFlagFilter.map((empFlagFilters) => {
            const { employee_id, firstName, lastName } = empFlagFilters;
            return (
              <li key={employee_id}>
                <EmployeesDetailWrapper>
                  <label>First name: </label>
                  <Span>{firstName}</Span>
                  <label>Last name: </label>
                  <Span>{lastName}</Span>
                </EmployeesDetailWrapper>
                <Button
                  onClick={() => {
                    dispatch({
                      type: ActionTypes.SET_FORM_STATE_FEEDBACK,
                      form: "save",
                      data: {},
                      id: employee_id,
                    });
                  }}
                >
                  Add Feedback
                </Button>
              </li>
            );
          })}
        </Unordered>
        <Feedback />
      </EmployeesWrapper>
    </div>
  );
};
