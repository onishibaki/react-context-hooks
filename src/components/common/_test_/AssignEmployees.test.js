import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { configure, mount } from "enzyme";

import { AssignEmployeeList } from "../AssignEmployees";
import { AppEmployeesProvider } from "../../reducer";
import { Feedback } from "../Feedback";

configure({ adapter: new Adapter() });

const employees = {
  status: 1,
  message: "Success",
  employees: [
    {
      employee_id: 1,
      firstName: "Mau",
      lastName: "Maya",
      flag: "1",
      feedbacklist: [],
    },
  ],
};

const NoAssignedEmployees = {
  status: 1,
  message: "Success",
  employees: [
    {
      employee_id: 1,
      firstName: "Felix",
      lastName: "San",
      flag: "0",
      feedbacklist: [],
    },
  ],
};

const databaseError = {
  status: 0,
  message: "Database Error",
  employees: [],
};

describe("Assigned Employees Component Test", () => {
  let wrapperEmployees;
  beforeEach(() => {
    wrapperEmployees = mount(
      <AppEmployeesProvider>
        <AssignEmployeeList props={employees} />
      </AppEmployeesProvider>
    );
  });

  afterEach(cleanup);
  it("Assigned Employees Component Renders Correctlyy", () => {
    const div = document.createElement("div");
    render(
      <AppEmployeesProvider>
        <AssignEmployeeList props={employees} />
      </AppEmployeesProvider>,
      div
    );
  });

  it("Check if No assigned Employees", () => {
    const wrapper = mount(
      <AppEmployeesProvider>
        <AssignEmployeeList props={NoAssignedEmployees} />
      </AppEmployeesProvider>
    );
    const noAssigned = wrapper.find("h5").at(0).text();
    expect(noAssigned).toEqual("No Assigned Employees");
  });

  it("Check if Assigned Employees", () => {
    const firstName = wrapperEmployees.find("span").at(1).text();
    const LastName = wrapperEmployees.find("span").at(2).text();
    expect(firstName).toEqual("Mau");
    expect(LastName).toEqual("Maya");
  });

  it("Check Database Error Display", () => {
    const wrapper = mount(
      <AppEmployeesProvider>
        <AssignEmployeeList props={databaseError} />
      </AppEmployeesProvider>
    );

    const errorDisplay = wrapper.find("div").text();
    expect(errorDisplay).toEqual("Database Error");
  });

  it("Check if FeedBack Component is Declare", () => {
    expect(wrapperEmployees.find(Feedback).length).toBe(1);
  });

  it("Check if FeedBack Component is Diplayed When Button Click", () => {
    wrapperEmployees.find("button").simulate("click");
    expect(wrapperEmployees.find(".feedback-container").exists()).toBeTruthy();
  });

  it("Assigned Employees Component Test Check Matches Snapshot Display Employee", () => {
    const Snapshot = renderer
      .create(
        <AppEmployeesProvider>
          <AssignEmployeeList props={employees} />
        </AppEmployeesProvider>
      )
      .toJSON();
    expect(Snapshot).toMatchSnapshot();
  });

  it("Assigned Employees Component Test Check Matches Snapshot Display Error", () => {
    const Snapshot = renderer
      .create(
        <AppEmployeesProvider>
          <AssignEmployeeList props={databaseError} />
        </AppEmployeesProvider>
      )
      .toJSON();
    expect(Snapshot).toMatchSnapshot();
  });
});
