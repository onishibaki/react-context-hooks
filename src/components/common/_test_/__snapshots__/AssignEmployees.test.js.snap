// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assigned Employees Component Test Assigned Employees Component Test Check Matches Snapshot Display Employee 1`] = `
<div
  className="main-container"
>
  <div
    className="sc-AxheI hszFvt"
  >
    <div
      className="sc-AxmLO fuyvoK"
    >
      <span>
        <a
          href="/"
        >
          Home
        </a>
      </span>
    </div>
    <ul
      className="sc-fzoLsD eURtDx"
    >
      <li>
        <div
          className="sc-Axmtr cRtglc"
        >
          <label>
            First name: 
          </label>
          <span
            className="sc-fzpans eJqXNW"
          >
            Mau
          </span>
          <label>
            Last name: 
          </label>
          <span
            className="sc-fzpans eJqXNW"
          >
            Maya
          </span>
        </div>
        <button
          className="sc-fzozJi hIzJCc"
          onClick={[Function]}
        >
          Add Feedback
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Assigned Employees Component Test Assigned Employees Component Test Check Matches Snapshot Display Error 1`] = `
<div
  className="main-container"
>
  Database Error
</div>
`;
